### ISSUE TITLE
Invalid detection of `dead_code` for const used as const generic argument during trait impl (regressed in 1.80.0)

### ISSUE THREAD

**Issue by datdenkikniet** [rust-lang/rust#128617]
In the following snippet, `CONST` gets a `dead_code` warning. However, `CONST` is most definitely used and the trait is also pub, so this warning seems incorrect.

Removing `CONST` causes the crate to no longer compile, so this warning seems to be incorrect.

```
pub struct Value<const NUMBER: u8> {}

pub trait Trait {}

const CONST: u8 = 11;

impl Trait for Value<CONST> {}
```
Rust version: rustc 1.80.0 (051478957 2024-07-21)

This code compiles without warnings on 1.79.0 (rustc 1.79.0 (129f3b996 2024-06-10))

**Comment by datdenkikniet** [comment]
Potentially fixed by https://github.com/rust-lang/rust/pull/128404

**Comment by compiler-errors** [comment]
@datdenkikniet: This is indeed fixed by that PR.

**Comment by compiler-errors** [comment]
cc @mu001999: This is another regression that should be fixed first if the dead code changes are re-landed.

**Comment by datdenkikniet** [comment]
This is fixed in the new point release, 1.80.1. Should we close this issue, or keep it open for tracking?

**Comment by jieyouxu** [comment]
This is probably worth keeping open for tracking purposes, and probably should be closed by the re-land PR(s) that add this case as a regression(?) test.

**Comment by jakubadamw** [comment]
This was fixed in 1.80.1 (https://github.com/rust-lang/rust/pull/128404) and seems to never have re-regressed. Might need a regression test, though.

@rustbot label +E-needs-test
